rr_iarrc/src/finish_line_watcher/finish_line_watcher.cpp

```diff
@@ -78,23 +74,85 @@ void ImageCB(const sensor_msgs::ImageConstPtr& msg) {
     frame = cv_ptr->image;
     blockEnvironment(frame);
 
-    auto count = countNonZero(frame);
+    // Do contour detection
+    vector<vector<Point>> contours;
+    vector<Vec4i> hierarchy;
+    Mat contourImg = frame.clone();
+    findContours(contourImg, contours, hierarchy, CV_RETR_LIST, CV_CHAIN_APPROX_SIMPLE);
+
+    Mat drawing = Mat::zeros(contourImg.size(), CV_8UC3);
+    double maxLength = 0.0;
+    double maxLengthYPos = 0.0;
+    for (size_t i = 0; i < contours.size(); i++) {
+        // Find the average slopes of the contour
+        Point start = contours.at(i).at(0);
+        Point end = contours.at(i).at((contours.at(i).size() - 1) / 2);
+        double rise = start.y - end.y;
+        double run = start.x - end.x + 0.01;
+        double slope = std::abs(rise / run);
+        // Also find the length
+        double length = arcLength(contours.at(i), false);
+
+        // Filter by slope and contour
+        if (slope > 0.0001 && slope < slope_cutoff && length > length_cutoff) {
+            // Draw for debug image
+            drawContours(drawing, contours, (int)i, Scalar(255, 0, 0), 2, LINE_8, hierarchy, 0);
+
+            if (length > maxLength) {
```

Comment:
Besides length, other shape factors like size could be important possibly. Note the old detector cared about the number of pixels.